src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,94 +10,183 @@ internal struct FileStatus
+                }
+                else
+                {
+                    // Others permissions
+                    readBit = Interop.Sys.Permissions.S_IROTH;
+                    writeBit = Interop.Sys.Permissions.S_IWOTH;
+                }
+#endif
+
+                return (_fileCache.Mode & (int)readBit)  != 0 && // has read permission
+                       (_fileCache.Mode & (int)writeBit) == 0;   // but not write permission
             }
-            else
+        }
+
+        // Checks if the main path is a symbolic link
+        // Only call if Refresh has been successfully called at least once
+        private bool HasSymbolicLinkFlag
+        {
+            get
             {
-                // Others permissions
-                readBit = Interop.Sys.Permissions.S_IROTH;
-                writeBit = Interop.Sys.Permissions.S_IWOTH;
+                Debug.Assert(IsFileCacheInitialized);
+                return (_fileCache.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFLNK;
             }
-#endif
+        }
+
+        // Checks if the specified cache (lstat=_fileCache, stat=_symlinkCache) has the directory attribute set
+        // Only call if Refresh has been successfully called at least once, and you're
+        // certain the passed-in cache was successfully retrieved
+        private bool CacheHasDirectoryFlag(Interop.Sys.FileStatus cache) => (cache.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR;
+
+        internal static void Initialize(ref FileStatus status, bool initiallyDirectory)
+        {
+            status.InitiallyDirectory = initiallyDirectory;
+            status.InvalidateCaches();
+        }
+
+        // Sets the cache initialization flags to -1, which means the caches are now uninitialized
+        internal void InvalidateCaches()
+        {
+            _initializedFileCache = -1;
+            _initializedSymlinkCache = -1;
+        }
 
-            return ((_fileStatus.Mode & (int)readBit) != 0 && // has read permission
-                (_fileStatus.Mode & (int)writeBit) == 0);     // but not write permission
+        internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            EnsureCachesInitialized(path, continueOnError);
+            return HasReadOnlyFlag;
         }
 
-        public FileAttributes GetAttributes(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName)
+        internal bool IsHidden(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName, bool continueOnError = false)
         {
-            // IMPORTANT: Attribute logic must match the logic in FileSystemEntry
+            EnsureCachesInitialized(path, continueOnError);
+            return IsNameHidden(fileName) || HasHiddenFlag;
+        }
 
-            EnsureStatInitialized(path);
+        internal bool IsNameHidden(ReadOnlySpan<char> fileName) => fileName.Length > 0 && fileName[0] == '.';
```

Comment:
It seems this more perf optimized:
```suggestion
        internal bool IsNameHidden(ReadOnlySpan<char> fileName) => fileName.StartsWith('.');
```

Reply:
Good point. I wrote a couple of quick benchmarks to compare their times:

```cs

private readonly string _myString = "MyString";

[Benchmark]
public bool FirstChar_StartsWith() => _myString.StartsWith('.');

[Benchmark]
public bool FirstChar_ManualCheck() => _myString.Length > 0 && _myString[0] == '.';
```

And `StartsWith` was faster:

|                Method |      Mean |     Error |    StdDev |    Median |       Min |       Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|------:|------:|----------:|
|  FirstChar_StartsWith | 0.0441 ns | 0.0056 ns | 0.0052 ns | 0.0423 ns | 0.0361 ns | 0.0551 ns |     - |     - |     - |         - |
| FirstChar_ManualCheck | 0.2870 ns | 0.0056 ns | 0.0049 ns | 0.2863 ns | 0.2791 ns | 0.2962 ns |     - |     - |     - |         - |

Reply:
I opened an issue to address this separately, so we don't block this PR (the last benchmarks looked fine):
https://github.com/dotnet/runtime/issues/52456

Reply:
Update: I just realized I originally tested using a `string`, not a `ReadOnlySpan<char>`. I tested it again. I realized I cannot pass a char as an argument to `ROS.StartsWith`, only a ROS or a string.

I tested it with a string and it does have a perf regression:

```cs
private readonly string _myString = "MyString";

[Benchmark]
public bool FirstChar_StartsWith()
{
    ReadOnlySpan<char> s = _myString;
    return s.StartsWith(".");
}

[Benchmark]
public bool FirstChar_ManualCheck()
{
    ReadOnlySpan<char> s = _myString;
    return s.Length > 0 && s[0] == '.';
}
```

|                Method |      Mean |     Error |    StdDev |    Median |       Min |       Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|------:|------:|----------:|
|  FirstChar_StartsWith | 3.0446 ns | 0.0192 ns | 0.0180 ns | 3.0483 ns | 3.0082 ns | 3.0689 ns |     - |     - |     - |         - |
| FirstChar_ManualCheck | 0.4466 ns | 0.0073 ns | 0.0069 ns | 0.4458 ns | 0.4338 ns | 0.4588 ns |     - |     - |     - |         - |

Reply:
Even with a cached dot in a char array, the benchmark is still regressed considerably:

```cs
private readonly string _myString = "MyString";
private readonly char[] _c = { '.' };

[Benchmark]
public bool FirstChar_StartsWith()
{
    ReadOnlySpan<char> s = _myString;
    return s.StartsWith(_c);
}

[Benchmark]
public bool FirstChar_ManualCheck()
{
    ReadOnlySpan<char> s = _myString;
    return s.Length > 0 && s[0] == '.';
}
```

|                Method |      Mean |     Error |    StdDev |    Median |       Min |       Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|------:|------:|----------:|
|  FirstChar_StartsWith | 3.1216 ns | 0.0155 ns | 0.0138 ns | 3.1266 ns | 3.0897 ns | 3.1414 ns |     - |     - |     - |         - |
| FirstChar_ManualCheck | 0.4491 ns | 0.0087 ns | 0.0077 ns | 0.4497 ns | 0.4345 ns | 0.4637 ns |     - |     - |     - |         - |